### PR TITLE
Feature: Add API for XPath parsing and processing (TEDEFO-2401)

### PR DESCRIPTION
**Pull request by bertrand-lorentz**
This consolidates the various XPath processing capabilities from the EFX toolkit into the eForms core library. See commit messages for more details.

This just adds APIs, so existing usage is not impacted. The removal of the corresponding features from the EFX toolkit is prepared and will be in a PR there once this is OK.

I'd welcome suggestions on the names of the new classes in the added API (XPathProcessor and XPathInfo).

### CODE REVIEW

File: src/main/java/eu/europa/ted/eforms/xpath/XPathInfo.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package eu.europa.ted.eforms.xpath;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class XPathInfo {
+	private LinkedList<XPathStep> steps = new LinkedList<>();
```
Comment (rouschr) [inline src/main/java/eu/europa/ted/eforms/xpath/XPathInfo.java]: Looks like this needs formatting

File: src/main/java/eu/europa/ted/eforms/xpath/XPathListenerImpl.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    }
+  }
+
+  /**
+   * Helper method that returns the input text that matched a parser rule context. It is useful
+   * because {@link ParserRuleContext#getText()} omits whitespace and other lexer tokens in the
+   * HIDDEN channel.
+   *
+   * @param context
+   * @return
```
Comment (rouschr) [inline src/main/java/eu/europa/ted/eforms/xpath/XPathListenerImpl.java]: This Java docs has empty param and return.

File: src/main/java/eu/europa/ted/eforms/xpath/XPathStep.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+public class XPathStep implements Comparable<XPathStep> {
+  private final String stepText;
+  private final List<String> predicates;
+
+  public XPathStep(String stepText, List<String> predicates) {
+		this.stepText = StringUtils.strip(stepText);
+    this.predicates = predicates;
+	}
```
Comment (rouschr) [inline src/main/java/eu/europa/ted/eforms/xpath/XPathStep.java]: Looks like this needs formatting

**Review by rouschr — CHANGES_REQUESTED** [review]
Apart from JavaDocs and formatting, I think as there are unit tests this should be at good start.

File: src/main/java/eu/europa/ted/eforms/xpath/XPathInfo.java (hunk 1 above)
Comment (bertrand-lorentz) [reply src/main/java/eu/europa/ted/eforms/xpath/XPathInfo.java]: VS Code keeps switching to indenting with tabs in new files, and a few tabs were left.
Fixed in added commit.

File: src/main/java/eu/europa/ted/eforms/xpath/XPathListenerImpl.java (hunk 2 above)
Comment (bertrand-lorentz) [reply src/main/java/eu/europa/ted/eforms/xpath/XPathListenerImpl.java]: This was copied from the code in EFX Toolkit.
Fixed in added commit.

File: src/main/java/eu/europa/ted/eforms/xpath/XPathStep.java (hunk 3 above)
Comment (bertrand-lorentz) [reply src/main/java/eu/europa/ted/eforms/xpath/XPathStep.java]: Fixed in added commit.

**Review by rouschr — APPROVED** [review]

**Review by rousso — APPROVED** [review]

**Pull request merged**